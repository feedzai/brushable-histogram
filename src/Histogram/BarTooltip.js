import React from "react";
import PropTypes from "prop-types";
import Portal from "../common/components/Portal";
import { HISTOGRAM_BAR_TOOLTIP_WIDTH, HISTOGRAM_BAR_TOOLTIP_HEIGHT } from "../constants";

/**
 * BarTootip
 *
 * Renders the histogram bar tooltip that allows to display information about a specific bar.
 * The tooltip content is customizable and is provided by the tooltipBarCustomization custom render function.
 *
 * @author Nuno Neves (nuno.neves@feedzai.com)
 */

const propsTypes = {
    selectedBarPosition: PropTypes.shape({
        top: PropTypes.number,
        left: PropTypes.number,
        width: PropTypes.number
    }),
    currentBar: PropTypes.object.isRequired,
    tooltipBarCustomization: PropTypes.func.isRequired
};
const defaultProps = {
    selectedBarPosition: {
        top: 0,
        left: 0,
        width: 0
    }
};

function BarTooltip({ selectedBarPosition, currentBar, tooltipBarCustomization }) {
    if (typeof tooltipBarCustomization !== "function") {
        return null;
    }

    const { top, left, width } = selectedBarPosition;
    const tooltipStyle = {
        position: "fixed",
        left: `${left - (HISTOGRAM_BAR_TOOLTIP_WIDTH / 2) + (width / 2)}px`,
        top: `${top - HISTOGRAM_BAR_TOOLTIP_HEIGHT}px`
    };

    const tooltipElement = tooltipBarCustomization(currentBar);

    return (
        <Portal>
            <div
                className="fdz-css-graph-histogram-bars--tooltip"
                style={tooltipStyle}
            >
                {tooltipElement}
            </div>
        </Portal>
    );
}

BarTooltip.propTypes = propsTypes;
BarTooltip.defaultProps = defaultProps;

export default React.memo(BarTooltip);
