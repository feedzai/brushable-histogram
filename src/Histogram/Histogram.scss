.fdz-css-graph-histogram {
    margin-top: 80px;

    $color-blue-500: #2196f3;
    $color-blue-300: #64b5f6;
    $color-blue-200: #90caf9;
    $color-blue-100: #bbdefb;

    $color-blue-gray-600: #546e7a;
    $color-blue-gray-500: #607d8b;
    $color-blue-gray-300: #90a4ae;
    $color-blue-gray-100: #cfd8dc;

    $grey-6: #bfbfbf;

    $tooltip-width: 100px;
    $tooltip-heigth: 40px;

    $histogram-tooltip-z-index: 11;

    $tooltip-arrow-height: 6px;

    &-bars{
        rect {
            fill: rgba($color-blue-500, 0.4);
        }
        rect:hover {
            @extend .fdz-css-cursor-pointer;
            fill: rgba($color-blue-300, 0.4);
        }

        &--tooltip {
            background-color: $color-blue-gray-600;
            width: $tooltip-width;
            padding: 0.3em 0.1em 0.3em 0.1em;
            margin-left: -$tooltip-width / 2;
            margin-top: -$tooltip-heigth - 10;
            text-align: center;
            z-index: $histogram-tooltip-z-index;
            font-family: 'Roboto Mono', Arial, sans-serif;

            &:after {
                top: 100%;
                left: 50%;
                border: solid transparent;
                content: " ";
                position: absolute;
                pointer-events: none;
                border-top-color: $color-blue-gray-600;
                border-width: $tooltip-arrow-height;
                margin-left: -$tooltip-arrow-height;
            }

            &-value {
                color: $color-blue-200;
                font-size: 10px;
                margin-bottom: 2px;
            }

            &-dates {
                color: $color-blue-gray-100;
                font-size: 8px;
            }
        }

    }

    &-axis-x {
        // d3v4 provides default styles and overrides previously defined font.
        font-family: 'Roboto Mono', Arial, sans-serif;
        font-size: 7pt;

        .domain{
            stroke:$color-blue-gray-300;
        }

        text {
            fill: $color-blue-gray-300;
            text-anchor: right;
        }

        line {
            stroke: $color-blue-gray-300;
        }
    }

    &-axis-y {
        // d3v4 provides default styles and overrides previously defined font.
        font-family: 'Roboto Mono', Arial, sans-serif;
        font-size: 7pt;

        .domain{
            stroke: transparent;
        }

        text {
            fill: $color-blue-gray-300;
            text-anchor: start;
        }

        line {
            stroke: $color-blue-gray-300;
        }
    }

    &-zoom {
        cursor: move;
        fill: none;
        pointer-events: all;
    }

    &-brush {
        rect {
            &.selection {
                fill: $color-blue-100;
                fill-opacity: 0.3;
                stroke-width: 0;
            }

            &.handle {
                fill: $color-blue-gray-500;
                width: 5px;
                stroke: $color-blue-gray-500;
                stroke-width: 1px;
            }
        }
    }

    &-density {
        &__canvas {
            position: relative;
        }

        &__wrapper {
            display: grid;
            grid-template-columns: auto auto;
        }
    }

    .fdz-css-cursor-pointer {
        cursor: pointer;
    }

    .fdz-css-play-btn {
        line-height: 1.5;
        padding: 0 15px;
        border-radius: 4px;
        user-select: none;
        transition: all 0.3s cubic-bezier(0.645, 0.045, 0.355, 1);
        position: relative;
        box-shadow: 0 2px 0 rgba(0, 0, 0, 0.015);
        cursor: pointer;
        top: -4px;


        background: transparent;
        border-color: transparent;
        color: $grey-6;
        font-size: 18px;

        &:hover, &:active, &:focus {
            background: transparent;
            border-color: transparent;
        }

        &:hover {
            color: #40a9ff;
        }

        .anticon {
            transition: margin-left 0.3s cubic-bezier(0.645, 0.045, 0.355, 1);
            line-height: 1;
            pointer-events: none;
            vertical-align: -0.125em;
            
            svg {
                line-height: 1;
            }
        }
    }
}