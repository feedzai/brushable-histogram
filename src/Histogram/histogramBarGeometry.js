/**
 * histogramBarGeometry
 *
 * This module contains the histogram bar position and dimension calculation.
 *
 * @author Luis Cardoso (luis.cardoso@feedzai.com)
 */

/**
  * Calculates the dimensions for the given `bar` given the scales and other parameters.
  * @returns {Object}
  */
export function calculateDimensions({ xScale, yScale, heightForBars, margin, bar }) {
    const width = xScale(bar.x1)
        - xScale(bar.x0) - margin;
    const height = heightForBars - yScale(bar.yValue);

    return { width, height };
}

/**
  * Calculates the position and dimensions for the given `bar` given the scales and other parameters.
  * @returns {Object}
  */
export function calculatePositionAndDimensions({ xScale, yScale, heightForBars, margin, bar }) {
    const { width, height } = calculateDimensions({ xScale, yScale, heightForBars, margin, bar });

    const x = xScale(bar.x0) + margin / 2;
    const y = yScale(bar.yValue);

    return {
        width,
        height,
        x,
        y
    };
}
