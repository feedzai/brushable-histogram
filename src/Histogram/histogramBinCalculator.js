import { histogram as d3Histogram } from "d3-array";

/**
 * histogramBinCalculator
 *
 * This module contains the histogram bin calculation logic.
 *
 * @author Luis Cardoso (luis.cardoso@feedzai.com)
 */

export default ({ xAccessor, yAccessor, histogramChartXScale, defaultBarCount, data }) => {
    // Setting the histogram function/converter
    const histogram = d3Histogram()
        .value(xAccessor)
        .domain(histogramChartXScale.domain()) // using the x-axis domain
        .thresholds(histogramChartXScale.ticks(defaultBarCount));

    // Calculating the time histogram bins
    return histogram(data).map((bar) => {
        const yValue = bar.reduce((sum, curr) => sum + yAccessor(curr), 0);

        return { ...bar, yValue };
    });
};
