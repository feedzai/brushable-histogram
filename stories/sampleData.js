const NUMBER_OF_POINTS = 60000;
const startTimestamp = 1533164400000;

/**
 * Calculates some sample data
 *
 * @param {number} numberOfPoints
 * @returns {Array.<Object>}
 */
export function calculate(numberOfPoints) {
    const data = [];

    for (let i = 0; i < numberOfPoints; i++) {
        data.push({
            "timestamp": startTimestamp + Math.abs(Math.floor(Math.cos(i) * 1000000000)),
            "total": Math.abs(Math.sin(i))
        });
    }

    return data;
}

export default calculate(NUMBER_OF_POINTS);

export const smallSample = calculate(10);
