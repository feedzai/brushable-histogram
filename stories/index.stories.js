import React, { Fragment } from "react";
import { timeFormat } from "d3-time-format";

import { storiesOf } from "@storybook/react";
import { withKnobs, number, object } from "@storybook/addon-knobs";

import sampleData from "./sampleData";
import Histogram from "../src/index";
import "../src/Histogram/Histogram.scss";

const stories = storiesOf("Histogram", module);

stories.addDecorator(withKnobs);

const formatDayFull = timeFormat("%e %b'%y");

function formatContextInterval(dateStart, dateEnd) {
    const formatStart = formatDayFull(dateStart);
    const formatEnd = formatDayFull(dateEnd);

    // If the formatted start and end date are the same then just display one
    if (formatStart === formatEnd) {
        return formatStart;
    }

    // Otherwise display the interval limits.
    return `${formatStart} - ${formatEnd}`;
}

function histogramTooltipBar(bar) {
    return (
        <Fragment>
            <div className="fdz-css-graph-histogram-bars--tooltip-value">
                {Math.floor(bar.yValue)} Events
            </div>
            <div className="fdz-css-graph-histogram-bars--tooltip-dates">
                {formatContextInterval(bar.x0, bar.x1)}
            </div>
        </Fragment>
    );
}

stories
    .add("Basic example", () => (
        <Histogram
            data={sampleData}
            xAccessor={(datapoint) => datapoint.timestamp}
            yAccessor={(datapoint) => datapoint.total}
        />
    ))
    .add("With a custom tooltip", () => (
        <Histogram
            data={sampleData}
            xAccessor={(datapoint) => datapoint.timestamp}
            yAccessor={(datapoint) => datapoint.total}
            tooltipBarCustomization={histogramTooltipBar}
        />
    ))
    .add("With a custom axis formatters", () => (
        <Histogram
            data={sampleData}
            xAccessor={(datapoint) => datapoint.timestamp}
            xAxisFormatter={formatDayFull}
            yAxisFormatter={(value) => (value > 0 ? `${value}$` : "")}
            yAccessor={(datapoint) => datapoint.total}
        />
    ))
    .add("Without the play button", () => (
        <Histogram
            data={sampleData}
            xAccessor={(datapoint) => datapoint.timestamp}
            yAccessor={(datapoint) => datapoint.total}
            renderPlayButton={false}
        />
    ))
    .add("With a custom a height", () => (
        <Histogram
            data={sampleData}
            xAccessor={(datapoint) => datapoint.timestamp}
            yAccessor={(datapoint) => datapoint.total}
            height={number("height (min 150)", 300)}
            tooltipBarCustomization={histogramTooltipBar}
        />
    ))
    .add("With no data", () => (
        <Histogram
            data={[]}
            xAccessor={(datapoint) => datapoint.timestamp}
            yAccessor={(datapoint) => datapoint.total}
            tooltipBarCustomization={histogramTooltipBar}
        />
    ))
    .add("Change data in order to update histogram", () => (
        <Histogram
            data={object("Data", [
                { timestamp: 1170070000000, total: 100 },
                { timestamp: 1270070000000, total: 23 },
                { timestamp: 1370070000000, total: 100 },
                { timestamp: 1470070000000, total: 23 },
                { timestamp: 1570070000000, total: 100 },
                { timestamp: 1670070000000, total: 23 },
                { timestamp: 1770070000000, total: 400 },
                { timestamp: 1870070000000, total: 200 }
            ])}
            xAccessor={(datapoint) => datapoint.timestamp}
            yAccessor={(datapoint) => datapoint.total}
            onIntervalChange={(domain) => console.log(domain)}
        />
    ));
