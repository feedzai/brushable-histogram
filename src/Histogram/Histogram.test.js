import React, { Fragment } from "react";
import { timeFormat } from "d3-time-format";
import { mount } from "enzyme";
import { Histogram } from "./Histogram";

import { smallSample } from "../../stories/sampleData";

jest.mock("../canvasRenderUtils");

// The calculation of the histogram bar positions can vary a bit
// depending on the system clock so we need to mock it to make sure
// we have no suprises in ci.
jest.mock("./histogramBinCalculator");

// The calcule of the bar positions and width depended a bit on the
// system clock. To avoid that dependency we mock the module that
// calculates those things.
jest.mock("./histogramBarGeometry");

const formatMinute = timeFormat("%I:%M");

function histogramYAxisFormatter(value) {
    if (value > 0 && Number.isInteger(value)) {
        return value;
    }
    return "";
}

function histogramTooltipBar(bar) {
    return (
        <Fragment>
            <div>
                {bar.yValue} Events
            </div>
            <div>
                {`${formatMinute(bar.x0)} - ${formatMinute(bar.x1)}`}
            </div>
        </Fragment>
    );
}

let wrapper, instance, onIntervalChangeSpy;

beforeEach(() => {
    onIntervalChangeSpy = jest.fn();

    wrapper = mount(<Histogram
        data={smallSample}
        size={{ width: 1000 }}
        height={150}
        xAccessor={(datapoint) => datapoint.timestamp}
        xAxisFormatter={formatMinute}
        yAccessor={(datapoint) => datapoint.total}
        yAxisFormatter={histogramYAxisFormatter}
        tooltipBarCustomization={histogramTooltipBar}
        onIntervalChange={onIntervalChangeSpy}
    />);
    instance = wrapper.instance();
});

describe("render", () => {
    it("does a baseline render", () => {
        expect(wrapper).toMatchSnapshot();
    });

    it("renders an empty chart if no data is passed", () => {
        jest.spyOn(Date, "now").mockImplementation(() => 1479427200000);

        const testWrapper = mount(<Histogram
            data={[]}
            size={{ width: 1000 }}
            height={150}
            xAccessor={(datapoint) => datapoint.timestamp}
            xAxisFormatter={formatMinute}
            yAccessor={(datapoint) => datapoint.total}
            yAxisFormatter={histogramYAxisFormatter}
            tooltipBarCustomization={histogramTooltipBar}
            onIntervalChange={onIntervalChangeSpy}
        />);

        expect(testWrapper).toMatchSnapshot();
    });
});

describe("_updateBrushedDomainAndReRenderTheHistogramPlot", () => {
    it("shouldn't call props.onIntervalChange if the domain didn't change", () => {
        instance._updateBrushedDomainAndReRenderTheHistogramPlot([
            new Date(1533309900034),
            new Date(1534164400001)
        ]);

        expect(onIntervalChangeSpy.mock.calls.length).toBe(0);
    });

    it("should call props.onIntervalChange if the domain did change", () => {
        instance._updateBrushedDomainAndReRenderTheHistogramPlot([
            new Date(1533309910034),
            new Date(1534164400001)
        ]);

        expect(onIntervalChangeSpy.mock.calls.length).toBe(1);
    });
});
