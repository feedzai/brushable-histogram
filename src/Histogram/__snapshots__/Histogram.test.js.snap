// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`_renderDensityChart should render an histogram bar 1`] = `
Array [
  <rect
    dataindex={0}
    height={10}
    onMouseEnter={[Function]}
    onMouseLeave={[Function]}
    width={10}
    x={1}
    y={1}
  />,
]
`;

exports[`render does a baseline render 1`] = `
<Histogram
  barOptions={
    Object {
      "margin": 1,
    }
  }
  data={
    Array [
      Object {
        "timestamp": 1534164400000,
        "total": 0,
      },
      Object {
        "timestamp": 1533704702305,
        "total": 0.8414709848078965,
      },
      Object {
        "timestamp": 1533580546837,
        "total": 0.9092974268256817,
      },
      Object {
        "timestamp": 1534154392497,
        "total": 0.1411200080598672,
      },
      Object {
        "timestamp": 1533818043621,
        "total": 0.7568024953079282,
      },
      Object {
        "timestamp": 1533448062185,
        "total": 0.9589242746631385,
      },
      Object {
        "timestamp": 1534124570286,
        "total": 0.27941549819892586,
      },
      Object {
        "timestamp": 1533918302254,
        "total": 0.6569865987187891,
      },
      Object {
        "timestamp": 1533309900034,
        "total": 0.9893582466233818,
      },
      Object {
        "timestamp": 1534075530262,
        "total": 0.4121184852417566,
      },
    ]
  }
  defaultBarCount={18}
  height={150}
  minZoomUnit={1000}
  onIntervalChange={[MockFunction]}
  padding={10}
  renderPlayButton={true}
  size={
    Object {
      "width": 1000,
    }
  }
  spaceBetweenCharts={10}
  tooltipBarCustomization={[Function]}
  xAccessor={[Function]}
  xAxisFormatter={[Function]}
  yAccessor={[Function]}
  yAxisFormatter={[Function]}
  yAxisTicks={3}
>
  <div
    className="fdz-css-graph-histogram"
  >
    <svg
      className="fdz-js-graph-histogram fdz-css-graph-histogram-chart"
      height={120}
      style={
        Object {
          "marginBottom": 10,
        }
      }
      width={1000}
    >
      <g
        className="fdz-css-graph-histogram-bars"
      >
        <rect
          dataindex={0}
          height={10}
          key="histogram-bin-1533294000000"
          onMouseEnter={[Function]}
          onMouseLeave={[Function]}
          width={10}
          x={1}
          y={1}
        />
        <rect
          dataindex={1}
          height={10}
          key="histogram-bin-1533337200000"
          onMouseEnter={[Function]}
          onMouseLeave={[Function]}
          width={10}
          x={1}
          y={1}
        />
        <rect
          dataindex={2}
          height={10}
          key="histogram-bin-1533380400000"
          onMouseEnter={[Function]}
          onMouseLeave={[Function]}
          width={10}
          x={1}
          y={1}
        />
        <rect
          dataindex={3}
          height={10}
          key="histogram-bin-1533423600000"
          onMouseEnter={[Function]}
          onMouseLeave={[Function]}
          width={10}
          x={1}
          y={1}
        />
        <rect
          dataindex={4}
          height={10}
          key="histogram-bin-1533466800000"
          onMouseEnter={[Function]}
          onMouseLeave={[Function]}
          width={10}
          x={1}
          y={1}
        />
        <rect
          dataindex={5}
          height={10}
          key="histogram-bin-1533510000000"
          onMouseEnter={[Function]}
          onMouseLeave={[Function]}
          width={10}
          x={1}
          y={1}
        />
        <rect
          dataindex={6}
          height={10}
          key="histogram-bin-1533553200000"
          onMouseEnter={[Function]}
          onMouseLeave={[Function]}
          width={10}
          x={1}
          y={1}
        />
        <rect
          dataindex={7}
          height={10}
          key="histogram-bin-1533596400000"
          onMouseEnter={[Function]}
          onMouseLeave={[Function]}
          width={10}
          x={1}
          y={1}
        />
        <rect
          dataindex={8}
          height={10}
          key="histogram-bin-1533639600000"
          onMouseEnter={[Function]}
          onMouseLeave={[Function]}
          width={10}
          x={1}
          y={1}
        />
        <rect
          dataindex={9}
          height={10}
          key="histogram-bin-1533682800000"
          onMouseEnter={[Function]}
          onMouseLeave={[Function]}
          width={10}
          x={1}
          y={1}
        />
        <rect
          dataindex={10}
          height={10}
          key="histogram-bin-1533726000000"
          onMouseEnter={[Function]}
          onMouseLeave={[Function]}
          width={10}
          x={1}
          y={1}
        />
        <rect
          dataindex={11}
          height={10}
          key="histogram-bin-1533769200000"
          onMouseEnter={[Function]}
          onMouseLeave={[Function]}
          width={10}
          x={1}
          y={1}
        />
        <rect
          dataindex={12}
          height={10}
          key="histogram-bin-1533812400000"
          onMouseEnter={[Function]}
          onMouseLeave={[Function]}
          width={10}
          x={1}
          y={1}
        />
        <rect
          dataindex={13}
          height={10}
          key="histogram-bin-1533855600000"
          onMouseEnter={[Function]}
          onMouseLeave={[Function]}
          width={10}
          x={1}
          y={1}
        />
        <rect
          dataindex={14}
          height={10}
          key="histogram-bin-1533898800000"
          onMouseEnter={[Function]}
          onMouseLeave={[Function]}
          width={10}
          x={1}
          y={1}
        />
        <rect
          dataindex={15}
          height={10}
          key="histogram-bin-1533942000000"
          onMouseEnter={[Function]}
          onMouseLeave={[Function]}
          width={10}
          x={1}
          y={1}
        />
        <rect
          dataindex={16}
          height={10}
          key="histogram-bin-1533985200000"
          onMouseEnter={[Function]}
          onMouseLeave={[Function]}
          width={10}
          x={1}
          y={1}
        />
        <rect
          dataindex={17}
          height={10}
          key="histogram-bin-1534028400000"
          onMouseEnter={[Function]}
          onMouseLeave={[Function]}
          width={10}
          x={1}
          y={1}
        />
        <rect
          dataindex={18}
          height={10}
          key="histogram-bin-1534071600000"
          onMouseEnter={[Function]}
          onMouseLeave={[Function]}
          width={10}
          x={1}
          y={1}
        />
        <rect
          dataindex={19}
          height={10}
          key="histogram-bin-1534114800000"
          onMouseEnter={[Function]}
          onMouseLeave={[Function]}
          width={10}
          x={1}
          y={1}
        />
        <rect
          dataindex={20}
          height={10}
          key="histogram-bin-1534158000000"
          onMouseEnter={[Function]}
          onMouseLeave={[Function]}
          width={10}
          x={1}
          y={1}
        />
        <rect
          dataindex={21}
          height={10}
          key="histogram-bin-1534201200000"
          onMouseEnter={[Function]}
          onMouseLeave={[Function]}
          width={10}
          x={1}
          y={1}
        />
      </g>
      <g
        className="fdz-js-graph-histogram-axis-x fdz-css-graph-histogram-axis-x"
        transform="translate(0, 102)"
      />
      <g
        className="fdz-js-graph-histogram-axis-y fdz-css-graph-histogram-axis-y"
        transform="translate(3, 3)"
      />
    </svg>
    <DensityChart
      brushDensityChartColor="rgba(33, 150, 243, 0.2)"
      brushDensityChartFadedColor="rgba(176, 190, 197, 0.2)"
      brushDomainMax={1534164400001}
      brushDomainMin={1533309900034}
      data={
        Array [
          Object {
            "timestamp": 1534164400000,
            "total": 0,
          },
          Object {
            "timestamp": 1533704702305,
            "total": 0.8414709848078965,
          },
          Object {
            "timestamp": 1533580546837,
            "total": 0.9092974268256817,
          },
          Object {
            "timestamp": 1534154392497,
            "total": 0.1411200080598672,
          },
          Object {
            "timestamp": 1533818043621,
            "total": 0.7568024953079282,
          },
          Object {
            "timestamp": 1533448062185,
            "total": 0.9589242746631385,
          },
          Object {
            "timestamp": 1534124570286,
            "total": 0.27941549819892586,
          },
          Object {
            "timestamp": 1533918302254,
            "total": 0.6569865987187891,
          },
          Object {
            "timestamp": 1533309900034,
            "total": 0.9893582466233818,
          },
          Object {
            "timestamp": 1534075530262,
            "total": 0.4121184852417566,
          },
        ]
      }
      densityChartXScale={[Function]}
      height={20}
      onDomainChanged={[Function]}
      padding={10}
      renderPlayButton={true}
      spaceBetweenCharts={10}
      width={940}
      xAccessor={[Function]}
    >
      <div
        className="fdz-css-graph-histogram-density__wrapper"
      >
        <PlayButton
          brushDomainMax={1534164400001}
          brushDomainMin={1533309900034}
          densityChartXScale={[Function]}
          frameDelay={500}
          frameStep={0.025}
          moveBrush={[Function]}
          width={940}
        >
          <button
            className="ant-btn fdz-css-play-btn ant-btn-icon-only"
            onClick={[Function]}
            type="button"
          >
            <i
              className="anticon anticon-play-circle"
            >
              <svg
                fill="currentColor"
                height="1em"
                viewBox="64 64 896 896"
                width="1em"
              >
                <path
                  d="M512 64C264.6 64 64 264.6 64 512s200.6 448 448 448 448-200.6 448-448S759.4 64 512 64zm0 820c-205.4 
    0-372-166.6-372-372s166.6-372 372-372 372 166.6 372 372-166.6 372-372 372z"
                />
                <path
                  d="M719.4 499.1l-296.1-215A15.9 15.9 0 0 0 398 297v430c0 13.1 14.8 20.5 25.3 12.9l296.1-215a15.9 
    15.9 0 0 0 0-25.8zm-257.6 134V390.9L628.5 512 461.8 633.1z"
                />
              </svg>
            </i>
          </button>
        </PlayButton>
        <div
          className="fdz-css-graph-histogram-density"
          style={
            Object {
              "position": "relative",
            }
          }
        >
          <canvas
            aria-label="Density Chart"
            className="fdz-css-graph-histogram-density__canvas"
            height={20}
            style={
              Object {
                "left": 0,
              }
            }
            width={940}
          />
          <svg
            alt="Density Chart Brush"
            className="fdz-css-graph-histogram-brush"
            height={20}
            style={
              Object {
                "left": 0,
                "position": "absolute",
                "top": 0,
              }
            }
            width={940}
          />
        </div>
      </div>
    </DensityChart>
  </div>
</Histogram>
`;

exports[`render renders an empty chart if no data is passed 1`] = `
<Histogram
  barOptions={
    Object {
      "margin": 1,
    }
  }
  data={Array []}
  defaultBarCount={18}
  height={150}
  minZoomUnit={1000}
  onIntervalChange={[MockFunction]}
  padding={10}
  renderPlayButton={true}
  size={
    Object {
      "width": 1000,
    }
  }
  spaceBetweenCharts={10}
  tooltipBarCustomization={[Function]}
  xAccessor={[Function]}
  xAxisFormatter={[Function]}
  yAccessor={[Function]}
  yAxisFormatter={[Function]}
  yAxisTicks={3}
>
  <div
    className="fdz-css-graph-histogram"
  >
    <svg
      className="fdz-js-graph-histogram fdz-css-graph-histogram-chart"
      height={120}
      style={
        Object {
          "marginBottom": 10,
        }
      }
      width={1000}
    >
      <g
        className="fdz-css-graph-histogram-bars"
      >
        <rect
          dataindex={0}
          height={10}
          key="histogram-bin-1533294000000"
          onMouseEnter={[Function]}
          onMouseLeave={[Function]}
          width={10}
          x={1}
          y={1}
        />
        <rect
          dataindex={1}
          height={10}
          key="histogram-bin-1533337200000"
          onMouseEnter={[Function]}
          onMouseLeave={[Function]}
          width={10}
          x={1}
          y={1}
        />
        <rect
          dataindex={2}
          height={10}
          key="histogram-bin-1533380400000"
          onMouseEnter={[Function]}
          onMouseLeave={[Function]}
          width={10}
          x={1}
          y={1}
        />
        <rect
          dataindex={3}
          height={10}
          key="histogram-bin-1533423600000"
          onMouseEnter={[Function]}
          onMouseLeave={[Function]}
          width={10}
          x={1}
          y={1}
        />
        <rect
          dataindex={4}
          height={10}
          key="histogram-bin-1533466800000"
          onMouseEnter={[Function]}
          onMouseLeave={[Function]}
          width={10}
          x={1}
          y={1}
        />
        <rect
          dataindex={5}
          height={10}
          key="histogram-bin-1533510000000"
          onMouseEnter={[Function]}
          onMouseLeave={[Function]}
          width={10}
          x={1}
          y={1}
        />
        <rect
          dataindex={6}
          height={10}
          key="histogram-bin-1533553200000"
          onMouseEnter={[Function]}
          onMouseLeave={[Function]}
          width={10}
          x={1}
          y={1}
        />
        <rect
          dataindex={7}
          height={10}
          key="histogram-bin-1533596400000"
          onMouseEnter={[Function]}
          onMouseLeave={[Function]}
          width={10}
          x={1}
          y={1}
        />
        <rect
          dataindex={8}
          height={10}
          key="histogram-bin-1533639600000"
          onMouseEnter={[Function]}
          onMouseLeave={[Function]}
          width={10}
          x={1}
          y={1}
        />
        <rect
          dataindex={9}
          height={10}
          key="histogram-bin-1533682800000"
          onMouseEnter={[Function]}
          onMouseLeave={[Function]}
          width={10}
          x={1}
          y={1}
        />
        <rect
          dataindex={10}
          height={10}
          key="histogram-bin-1533726000000"
          onMouseEnter={[Function]}
          onMouseLeave={[Function]}
          width={10}
          x={1}
          y={1}
        />
        <rect
          dataindex={11}
          height={10}
          key="histogram-bin-1533769200000"
          onMouseEnter={[Function]}
          onMouseLeave={[Function]}
          width={10}
          x={1}
          y={1}
        />
        <rect
          dataindex={12}
          height={10}
          key="histogram-bin-1533812400000"
          onMouseEnter={[Function]}
          onMouseLeave={[Function]}
          width={10}
          x={1}
          y={1}
        />
        <rect
          dataindex={13}
          height={10}
          key="histogram-bin-1533855600000"
          onMouseEnter={[Function]}
          onMouseLeave={[Function]}
          width={10}
          x={1}
          y={1}
        />
        <rect
          dataindex={14}
          height={10}
          key="histogram-bin-1533898800000"
          onMouseEnter={[Function]}
          onMouseLeave={[Function]}
          width={10}
          x={1}
          y={1}
        />
        <rect
          dataindex={15}
          height={10}
          key="histogram-bin-1533942000000"
          onMouseEnter={[Function]}
          onMouseLeave={[Function]}
          width={10}
          x={1}
          y={1}
        />
        <rect
          dataindex={16}
          height={10}
          key="histogram-bin-1533985200000"
          onMouseEnter={[Function]}
          onMouseLeave={[Function]}
          width={10}
          x={1}
          y={1}
        />
        <rect
          dataindex={17}
          height={10}
          key="histogram-bin-1534028400000"
          onMouseEnter={[Function]}
          onMouseLeave={[Function]}
          width={10}
          x={1}
          y={1}
        />
        <rect
          dataindex={18}
          height={10}
          key="histogram-bin-1534071600000"
          onMouseEnter={[Function]}
          onMouseLeave={[Function]}
          width={10}
          x={1}
          y={1}
        />
        <rect
          dataindex={19}
          height={10}
          key="histogram-bin-1534114800000"
          onMouseEnter={[Function]}
          onMouseLeave={[Function]}
          width={10}
          x={1}
          y={1}
        />
        <rect
          dataindex={20}
          height={10}
          key="histogram-bin-1534158000000"
          onMouseEnter={[Function]}
          onMouseLeave={[Function]}
          width={10}
          x={1}
          y={1}
        />
        <rect
          dataindex={21}
          height={10}
          key="histogram-bin-1534201200000"
          onMouseEnter={[Function]}
          onMouseLeave={[Function]}
          width={10}
          x={1}
          y={1}
        />
      </g>
      <g
        className="fdz-js-graph-histogram-axis-x fdz-css-graph-histogram-axis-x"
        transform="translate(0, 102)"
      />
      <g
        className="fdz-js-graph-histogram-axis-y fdz-css-graph-histogram-axis-y"
        transform="translate(3, 3)"
      />
    </svg>
    <DensityChart
      brushDensityChartColor="rgba(33, 150, 243, 0.2)"
      brushDensityChartFadedColor="rgba(176, 190, 197, 0.2)"
      brushDomainMax={1479427200000}
      brushDomainMin={1479427200000}
      data={Array []}
      densityChartXScale={[Function]}
      height={20}
      onDomainChanged={[Function]}
      padding={10}
      renderPlayButton={false}
      spaceBetweenCharts={10}
      width={940}
      xAccessor={[Function]}
    >
      <div
        className="fdz-css-graph-histogram-density__wrapper"
      >
        <div
          className="fdz-css-graph-histogram-density"
          style={
            Object {
              "position": "relative",
            }
          }
        >
          <canvas
            aria-label="Density Chart"
            className="fdz-css-graph-histogram-density__canvas"
            height={20}
            style={
              Object {
                "left": 20,
              }
            }
            width={940}
          />
          <svg
            alt="Density Chart Brush"
            className="fdz-css-graph-histogram-brush"
            height={20}
            style={
              Object {
                "left": 20,
                "position": "absolute",
                "top": 0,
              }
            }
            width={940}
          />
        </div>
      </div>
    </DensityChart>
  </div>
</Histogram>
`;
